sh/drive-utils.sh

```diff
@@ -0,0 +1,448 @@
+#   Variables - API_URL, API_VERSION
+#   Functions - _url_encode, _json_value
+# Arguments: 3
+#   ${1} = dir name
+#   ${2} = root dir id of given dir
+#   ${3} = Access Token
+# Result: print folder id
+# Reference:
+#   https://developers.google.com/drive/api/v3/folder
+###################################################
+_create_directory() (
+    [ $# -lt 3 ] && printf "Missing arguments\n" && return 1
+    dirname_create_directory="${1##*/}" rootdir_create_directory="${2}" token_create_directory="${3}"
+    unset query_create_directory search_response_create_directory folder_id_create_directory
+
+    query_create_directory="$(_url_encode "mimeType='application/vnd.google-apps.folder' and name='${dirname_create_directory}' and trashed=false and '${rootdir_create_directory}' in parents")"
+
+    search_response_create_directory="$(curl --compressed -s \
+        -H "Authorization: Bearer ${token_create_directory}" \
+        "${API_URL}/drive/${API_VERSION}/files?q=${query_create_directory}&fields=files(id)&supportsAllDrives=true")" || :
+
+    folder_id_create_directory="$(printf "%s\n" "${search_response_create_directory}" | _json_value id 1 1)"
+
+    if [ -z "${folder_id_create_directory}" ]; then
+        unset create_folder_post_data_create_directory create_folder_response_create_directory
+        create_folder_post_data_create_directory="{\"mimeType\": \"application/vnd.google-apps.folder\",\"name\": \"${dirname_create_directory}\",\"parents\": [\"${rootdir_create_directory}\"]}"
+        create_folder_response_create_directory="$(curl --compressed -s \
+            -X POST \
+            -H "Authorization: Bearer ${token_create_directory}" \
+            -H "Content-Type: application/json; charset=UTF-8" \
+            -d "${create_folder_post_data_create_directory}" \
+            "${API_URL}/drive/${API_VERSION}/files?fields=id&supportsAllDrives=true")" || :
+        folder_id_create_directory="$(printf "%s\n" "${create_folder_response_create_directory}" | _json_value id 1 1)"
+    fi
+    { [ -z "${folder_id_create_directory}" ] && printf "%s\n" "${create_folder_response_create_directory}" 1>&2 && return 1; } || {
+        printf "%s\n" "${folder_id_create_directory}"
+    }
+    return 0
+)
+
+###################################################
+# Sub functions for _upload_file function - Start
+# generate resumable upload link
+_generate_upload_link() {
+    uploadlink_upload_file="$(curl --compressed -s \
+        -X "${request_method_upload_file}" \
+        -H "Authorization: Bearer ${token_upload_file}" \
+        -H "Content-Type: application/json; charset=UTF-8" \
+        -H "X-Upload-Content-Type: ${mime_type_upload_file}" \
+        -H "X-Upload-Content-Length: ${inputsize_upload_file}" \
+        -d "$postdata_upload_file" \
+        "${url_upload_file}" \
+        -D -)" || :
+    uploadlink_upload_file="$(printf "%s\n" "${uploadlink_upload_file}" | sed -n 's/[L,l]ocation: //p' | sed 's/'$'\r''//g')"
```

Comment:
Codacy found an issue: [In POSIX sh, $'..' is undefined.](https://app.codacy.com/manual/labbots/google-drive-upload/pullRequest?prid=5865418)